### DISCUSSION TITLE
Continuation Loop for Composer Version could use a size slider <403

### DISCUSSION THREAD

**Discussion by Timzoid** [asigalov61/Euterpe-X#2] [Ideas]
Well, I tried Euterpe this morning, both the fast and slow versions. It is hard for me to tell what is different from LAMC. The fast version seems to generate faster.  

Alex, you wrote that the size was fixed at around 400 for the output of the continuation in the composer versions, because that was the max in MuseNet, and a pretty good size. However, I varied my tokens size in MuseNet frequently, and used the 50, for the control it would offer. 

As I used MuseNet for about 5 hours a day, at least, for over three months, I gradually found that a setting of 150 tokens was the right length for me as a base length, the length that worked the best and still gave me creative choices, but it was handy to occasionally use 400 when I wanted to see where the program would take it, even if I didn't use that sequence in the final piece. 

So, I liked the flexibility of using different sizes, as I got a feel for how it was handling the particular music that it was generating. I never did a piece that was all 400 and used a 400 sequence only about once or twice, in any given 3 to 4 minute piece I built. Most of the time I stuck to the 150-250 range. 

I like the "Outro" feature in this. In over  90% of MuseNet compositions I wrote my own endings. However, the 3 minute mark in a MuseNet piece was reached, it had an uncanny way of producing some choices that were good, in part, for an ending. Still, this is really nice, having that feature.

One thing I liked about MuseNet, I worked on as many as 7 pieces at once, each open in a different tab, using the same intro, and then from the 4 or so which came out best, I'd pick the parts from each that went into a final piece. I did that only on pieces with an intro that worked really well. MuseNet was so simple to use, on the surface, but there were ways to work with it that were versatile, could increase the chances of producing something good.

**Comment by asigalov61** [comment]
@Timzoid I apologize for the delayed response...

Thank you for trying Euterpe! :)

Euterpe Large is my most capable model, although it does not have velocities as LAMC does.

Yes, it is very easy to add a slider to control the generated block size in Composer versions (both Euterpe and LAMC). I will consider adding this option when I post a new update.

And yes, you are correct. 400 tokens may be too much for some compositions and 150 tokens(as you have suggested) would probably work better. Less is better indeed as it indeed allows more control over the generated music.

Now, Outro option in Euterpe does not work too well because I think I screwed it up by setting it at 200 notes(600) tokens. That is too much for the model to handle (from my testing). But sometimes it does work well so I am glad you checked it out too.

I have another model/implementation that you might want to consider:
https://github.com/asigalov61/Allegro-Music-Transformer

This implementation uses MuseNet approach to generate music/continuations and it is similar in behavior to musenet because it uses chords counters and outro tokens together with the asynchrous encoding (similar to MuseNet). Especially the outro function works well in this implementation.

Check it out and let me know what you think.

And thank you again for your thoughts and feedback.

Alex.

**Comment by asigalov61** [comment]
@Timzoid I've done what you have suggested and added continuation blocks size controls and preview controls to the LAMC. Will add the same to Euterpe X shortly.

Closing this discussion for now as it is complete I think :)

And again, thank you for your feedback. It was useful.
